src/server/Adaptive.ReactiveTrader.Server.Pricing/PriceSource.cs

```diff
@@ -1,92 +1,120 @@
-            CreatePriceGenerator("AUDJPY", 88.666m, 3),
-            CreatePriceGenerator("AUDNZD", 1.08334m, 5),
-            CreatePriceGenerator("CADJPY", 89.7685m, 3),
-            CreatePriceGenerator("CHFUSD", 1.01027m, 5),
-            CreatePriceGenerator("EURNOK", 9.44156m, 4),
-            CreatePriceGenerator("EURSEK", 9.26876m, 4)
-        }).ToDictionary(x => x.Symbol);
-
       var keys = _priceGenerators.Keys.ToArray();
-
-      _timer = Observable.Interval(TimeSpan.FromMilliseconds(50)).Publish();
-      _disposable.Add(_timer.Subscribe(observer =>
+      _priceWalkTimer = new Timer();
+      _priceWalkTimer.Interval = 50;
+      _priceWalkTimer.Elapsed += delegate (object sender, ElapsedEventArgs e) { _priceGenerators[keys[Random.Next(_priceGenerators.Count)]].UpdateWalkPrice(); };
+      _disposables.Add(_priceWalkTimer);
+
+      _marketRatesTimer = new Timer();
+      _marketRatesTimer.Interval = 1000;
+      _marketRatesTimer.Elapsed += async delegate (object sender, ElapsedEventArgs e)
       {
-        if ((DateTime.UtcNow - _lastMarketUpdate).TotalHours > 1)
+        var now = GetThisHour();
+        if ((now - _lastMarketUpdate).TotalHours > HourlyMarketRateUpdateFrequency)
         {
-          //Just refresh the currency pairs from the markets once every hour.
-          RefreshMarketRates(DateTime.UtcNow);
+          try
+          {
+            _marketRatesTimer.Enabled = false;
+            await RefreshMarketRates(now);
+            Log.Information("Market Rates update succeeded");
+          }
+          catch (Exception ex)
+          {
+            Log.Error(ex, "Market Rates update failed with exception");
+          }
+          finally
+          {
+            _marketRatesTimer.Enabled = true;
+          }
         }
-        _priceGenerators[keys[Random.Next(_priceGenerators.Count)]].UpdateWalkPrice();
-      }));
+      };
+      _disposables.Add(_marketRatesTimer);
 
-      _timer.Connect();
+      _priceWalkTimer.Enabled = true;
+      _marketRatesTimer.Enabled = true;
     }
 
     public void Dispose()
     {
-      _disposable.Dispose();
+      _disposables.Dispose();
     }
 
-    private static IPriceGenerator CreatePriceGenerator(string symbol, decimal initial, int precision)
+    private static IPriceGenerator CreatePriceGenerator(string baseCcy, string quoteCcy, decimal initial, int precision)
     {
-      return new MeanReversionRandomWalkPriceGenerator(symbol, initial, precision);
+      return new MeanReversionRandomWalkPriceGenerator(new CurrencyPair(baseCcy, quoteCcy), initial, precision);
     }
 
-    private void RefreshMarketRates(DateTime refreshDateTime)
+    private async Task RefreshMarketRates(DateTime refreshDateTime)
     {
       foreach (var adapter in _marketAdapters)
       {
-        var marketData = new List<MarketData>(adapter.GetMarketData().Result);
+        var marketData = new List<MarketData>(await adapter.GetMarketData());
```

Comment:
Is this going to do what we want if this GetMarketData method throws?
I'd have thought in that case we'd want to fall back to one of the other providers?

Reply:
Ok, I was a bit lazy and assumed that all adapters, now and in the future would trap their own exceptions. I have added a lovely try{} catch{} here to protect future adapters.